zerver/tests/test_message_fetch.py

```diff
@@ -3442,6 +3444,96 @@ def test_parse_anchor_value(self) -> None:
         result = orjson.loads(payload.content)
         self.assertEqual(result["anchor"], LARGER_THAN_MAX_MESSAGE_ID)
 
+    def test_narrow_for_anchor_date(self) -> None:
+        user_profile = self.example_user("hamlet")
+        stream = self.make_stream("date-anchor")
+        self.subscribe(user_profile, stream.name)
+        current_datetime = datetime.datetime.now(tz=datetime.timezone.utc)
+
+        msg_ids = []
+        with time_machine.travel(current_datetime, tick=False) as traveller:
+            for _ in range(20):
+                msg_id = self.send_stream_message(user_profile, stream.name)
+                msg_ids.append(msg_id)
+                traveller.shift(datetime.timedelta(days=1))
+
+        search_by_days = 10
+        search_datetime = current_datetime + datetime.timedelta(days=search_by_days)
+        search_date_str = search_datetime.isoformat()
+        query_params = dict(
+            anchor="date",
+            anchor_date=search_date_str,
+            num_before=5,
+            num_after=5,
+            narrow=f'[["stream", "{stream.name}"]]',
+        )
+
+        request = HostRequestMock(query_params, user_profile)
+        payload = get_messages_backend(request, user_profile)
+        result = orjson.loads(payload.content)
+        self.assertEqual(result["anchor"], msg_ids[search_by_days])
+
+        search_datetime = current_datetime - datetime.timedelta(days=search_by_days * 4)
+        search_date_str = search_datetime.isoformat()
+        query_params = dict(
+            anchor="date",
+            anchor_date=search_date_str,
+            num_before=5,
+            num_after=5,
+            narrow=f'[["stream", "{stream.name}"]]',
+        )
+
+        request = HostRequestMock(query_params, user_profile)
+        payload = get_messages_backend(request, user_profile)
+        result = orjson.loads(payload.content)
+        self.assertEqual(result["anchor"], msg_ids[0])
+
+        # case search_datetime > datetime of latest message sent
```

Comment:
I'd like to see this test improved to more clearly verify all the corner cases in the actual logic, with more comments like this one, and also checking which message IDs were returned, not just the anchor.

I would also like to see a case using the date-only ISO format.

Reply:
Added more test cases.